[Feat] Streamlit 앱에 실험 데이터 분석을 위한 기능 추가

gsgh3016 opened the pull request:
## 📝 Summary

<!--
PR의 주요 내용을 간단히 요약해주세요.
예: 로그인 기능에서 사용자 세션 관리 개선.
-->
데이터 관련 실험 시, 새로운 열이 생길 가능성이 높습니다.
현재 Streamlit 앱의 `데이터 탐색 > 문제 형태로 확인` 기능을 바탕으로 새로운 탭과 레이아웃을 추가합니다.
실험 데이터를 위한 파일 업로더, `실험 데이터` 탭, `실험 데이터 > 문제 형태로 확인` 기능을 추가했습니다.

## ✅ Checklist

<!--
해당되는 항목을 [x]로 만들어주세요.
예: - [x] 관련 이슈가 명시되어 있습니다.

해당되지 않는 항목은 앞뒤로 ~를 붙여서 취소선을 그어주세요.
예: - ~[ ] 관련 이슈가 명시되어 있습니다.~
-->

- [X] 관련 이슈가 명시되어 있습니다.
- [X] 테스트가 완료되었습니다.
- ~[ ] 문서 업데이트가 포함되었습니다.~
- [X] 코드 리뷰를 위한 사전 검토를 완료했습니다.

## 📄 Description

<!--
PR의 변경 사항을 구체적으로 설명해주세요.
예: 이번 변경 사항에는 로그인 세션 시간 연장 및 오류 메시지 개선이 포함됩니다.
-->

### 1. 데이터 경로
기본 데이터 로드와 유사하게, 데이터 경로를 환경 변수로 관리합니다.
```bash
# .env
STREAMLIT_EXPERIMENT_DATA_PATH={your_experiment_result_path}
```

```python
# analysis_dashboard.py
import os
from dotenv import load_dotenv
...
load_dotenv()
...
exp_df = pd.read_csv(os.getenv("STREAMLIT_EXPERIMENT_DATA_PATH"))
```

### 2. 새로운 파일 업로더 및 탭 생성
![image](https://github.com/user-attachments/assets/ccf04404-d2b8-44fa-b304-858138bf53f6)

기본 학습 데이터를 `df` 변수에 저장, 실험 데이터를 `exp_df` 변수에 저장하여 따로 관리합니다.
따라서 새로운 파일 업로더와 탭 기능을 추가합니다.

### 3. ~`key_manager` 객체 생성~ &rarr; 탭 별로 key 값을 따로 관리하도록 수정
탭 별로 고유한 `Element ID`를 생성하도록 탭 이름을 인자로 렌더링 함수에 넘겨줬습니다.
추후 Streamlit 앱 기능 추가 및 재사용 시 `StreamlitDuplicateElementKey` 에러를 방지할 수 있습니다. 

### 4. 렌더링 모듈화
`display_data_tab` 함수로 기존 `tab2` 렌더링 구조를 모듈화하여 새로운 탭인 `tab4`에서 재사용하도록 코드를 수정합니다.

### 5. 전체 열을 순회하면서 새로운 열 렌더링
`실험 데이터 탐색 > 문제 형태로 확인`
![문제 형태로 확인](https://github.com/user-attachments/assets/43c1935f-2dd7-4c75-91ab-f2a46180eaca)

실험 데이터 칼럼 이름 정보를 전부 순회하면서 기본 데이터 칼럼에 없을 경우, 렌더링하는 기능을 추가했습니다.

## 💡 Notice (Optional)

<!--
리뷰어가 알아야 할 추가적인 사항이나 주의점이 있다면 작성해주세요.
예: 이 기능은 인증 모듈과 호환성을 고려해야 합니다.
-->
해당 기능이 제대로 동작하기 위해서는 데이터 저장 및 환경 변수 세팅이 필요합니다.
Description의 1. 데이터 경로에 세팅 방법을 작성했습니다.

Streamlit 앱 기능과 관련 없지만, Streamlit 기능 개발에 용이한 vscode 설정 파일을 추가했습니다.
analysis_dashboard.py에서 코드 라인에 breakpoint를 잡고 vscode 상단에 디버깅을 선택합니다.
<img width="395" alt="스크린샷 2024-11-23 오후 2 55 01" src="https://github.com/user-attachments/assets/bd4ff6a1-c3bd-4e57-9dee-046a1b42caca">

## 🔗 Related Issue(s)

<!--
이 PR과 관련된 이슈나 기존 PR이 있다면 번호를 언급하고, 필요 시 close할 이슈도 명시해주세요.
예: 관련 이슈 - #33, close #27
-->
close #39

gsgh3016 committed:
feat: Streamlit 탭 추가

gsgh3016 committed:
feat: `Element ID` 생성하는 `KeyManager` 클래스 추가

gsgh3016 committed:
chore: 실험 데이터 경로 세팅

gsgh3016 committed:
feat: 실험 데이터 탭에서 레이아웃 적용

gsgh3016 committed:
feat: 추가된 열 렌더링 기능 구현

eyeol reviewed (changes requested):
작업하느라 고생하셨습니다. 코멘트 확인 부탁드립니다.

gsgh3016 committed:
fix: 필요 없는 Element ID 값 제거

gsgh3016 committed:
chore: 환경 변수 오타 수정

gsgh3016 committed:
docs: 데이터 로드 주석 수정

gsgh3016 committed:
chore: Streamlit 디버깅 용 VScode 설정

jagaldol reviewed (commented):
streamlit과 그렇게 친숙하지 않아서 코드를 봐도 어렵네요

우리 중 @eyeol 님이 가장 잘 알거 같아서 @eyeol 에게 approve를 맡기겠습니다!

제가 보기엔 일단 이상이 없어 보입니다! 👀

gsgh3016 committed:
fix: `st.number_input`이 동작하지 않던 현상 해결

- Element ID가 달라지면서 세션에 저장됐던 변수 값을 읽지 못하는 것 확인
- key_manager 관련 코드 삭제
- 탭 별로 렌더링 모듈화 시, 탭 이름을 인자로 넣도록 변경

gsgh3016 committed:
fix: 실험 데이터 칼럼 검색 기능 수정

gsgh3016 committed:
refactor: key_manager 기능 및 파일 삭제

eyeol reviewed (approved):
수정 내용 확인했습니다. 고생하셨습니다.
